### PR TITLE
(800) Add role based access and apply to refer

### PR DESCRIPTION
**Pull request by jsrobertson**
## Context

We need to restrict access to anything referral related to users with the `ACP_REFERRER` role/authority.

https://trello.com/c/2K0xmD3j/800-restrict-being-able-to-make-a-referral-to-only-those-with-accredited-programme-referrer-permissions-ui-api

## Changes in this PR
Created a new `roleBasedAccessMiddleware` which accepts an options parameter with an `allowedRoles` Array.  We can then use this in each of the routes to restrict access to routes to users with only the specified, allowed roles.

Also updates res.locals.user to include an `hasReferrerRole` property which can be used in the templates.

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes API for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [ ] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

**Review by Gweaton — APPROVED** [review]
This looks good to me.

I'm approving so as not to block you, but before merging, could you add an integration test to verify that this redirects users to the `/authError` page when they don't have the right role, just for extra safety?

Also, let's hold off on merging until we've got those new users in.

File: integration_tests/mockApis/auth.ts
Comment (yndajas) [inline integration_tests/mockApis/auth.ts:12]: I'm not quite familiar with this syntax. Is it saying destructure `authorities` if passed in an object argument, otherwise set it to an empty array?

Then on line 103, destructure `authorities` from the object argument even if undefined?

Perhaps this is just stretching the limits of my TypeScript knowledge(!), but would this not work the same?

```ts
// L12
const createToken = (authorities: Array<ApplicationRoles> = []) => {

// L103
const token = (authorities?: Array<ApplicationRoles>) =>

// L115
        access_token: createToken(authorities),

// L167-169
    authorities = []

// L176
      token(authorities),
```

File: server/@types/server/index.d.ts
Comment (jsrobertson) [reply server/@types/server/index.d.ts:8]: Of course cheers! Meant to come back round and do this as I did have the ApplicationRoles being exported out of it at one point.

File: server/middleware/roleBasedAccessMiddleware.ts
Comment (jsrobertson) [reply server/middleware/roleBasedAccessMiddleware.ts:1]: Happy to add this now.  My initial thinking was add the disable comment once and then add in future if we needed to add more elsewhere.

File: server/middleware/populateCurrentUser.test.ts
Comment (jsrobertson) [reply server/middleware/populateCurrentUser.test.ts:46]: Aye

File: integration_tests/mockApis/auth.ts
Comment (jsrobertson) [reply integration_tests/mockApis/auth.ts:12]: Yeah pretty much, I was thinking ahead really incase we needed to stub any other values on the user for future tests.  Might be a bit premature.

Comment (yndajas) [reply integration_tests/mockApis/auth.ts:12]: Makes sense - I'm not close enough to this work to know how likely that is!

**Pull request merged**
